h/models/group.py

```diff
@@ -221,56 +222,67 @@ def is_public(self):
     def __acl__(self):
         terms = []
 
-        join_principal = _join_principal(self)
-        if join_principal is not None:
-            terms.append((security.Allow, join_principal, Permission.Group.JOIN))
+        # This principal is given to clients which log in using an OAuth client
+        # and secret to a particular authority
+        client_authority_principal = "client_authority:{}".format(self.authority)
+        # Given to logged in users to the authority they belong in
+        in_authority_principal = "authority:{}".format(self.authority)
+        # Logged in users are given group principals for all of their groups
+        in_group_principal = "group:{}".format(self.pubid)
 
-        read_principal = _read_principal(self)
-        if read_principal is not None:
-            terms.append((security.Allow, read_principal, Permission.Group.READ))
-            # Any user who can read the group should also be able to see
-            # who is a member of the group
-            terms.append((security.Allow, read_principal, Permission.Group.MEMBER_READ))
+        # General permissions ---------------------------------------------- #
 
-        flag_principal = _flag_principal(self)
-        if flag_principal is not None:
-            terms.append((security.Allow, flag_principal, Permission.Group.FLAG))
+        if self.joinable_by == JoinableBy.authority:
+            terms.append((Allow, in_authority_principal, Permission.Group.JOIN))
 
-        write_principal = _write_principal(self)
-        if write_principal is not None:
-            terms.append((security.Allow, write_principal, Permission.Group.WRITE))
```

Comment:
I'll use this one as an example of how these were refactored (they are all the same). First the function was moved inline so it looked like this:

```python
        def _write_principal(group):
            return {
                WriteableBy.authority: "authority:{}".format(group.authority),
                WriteableBy.members: "group:{}".format(group.pubid),
            }.get(group.writeable_by)

        write_principal = _write_principal(self)
        if write_principal is not None:
            terms.append((security.Allow, write_principal, Permission.Group.WRITE))
```

Then the logic this actually meant was extracted like this:

```python
        if self.writeable_by == WriteableBy.authority:
            terms.append((Allow, "authority:{}".format(self.authority), Permission.Group.WRITE))
        elif self.writeable_by == WriteableBy.members:
            terms.append((Allow,  "group:{}".format(self.pubid), Permission.Group.WRITE))
```

Finally the repeated principals were extracted as variables at the top resulting in:

```python
        if self.writeable_by == WriteableBy.authority:
            terms.append((Allow, in_authority_principal, Permission.Group.WRITE))
        elif self.writeable_by == WriteableBy.members:
            terms.append((Allow, in_group_principal, Permission.Group.WRITE))
```